### PR TITLE
Enable BDW topology

### PR DESCRIPTION
**Pull request by xiulipan**
Fix sof-bdw-rt286.tplg load
Add HSW/BDW nocodec topology.

### CODE REVIEW

**Comment by jajanusz** [comment]
@xiulipan Is it fix for CI?

**Comment by xiulipan** [comment]
@jajanusz This is enable for BDW platform, we are trying to add BDW into CI. But BDW can not pass manually test now

**Comment by lgirdwood** [comment]
@zrombel any reason why internal CI builds would fail on unrelated patch ? (Travis build passed)

**Comment by zrombel** [comment]
@lgirdwood looks like QuickBuild issue. Rerunning build.

File: tools/topology/sof-bdw-codec.m4 [hunk 1]
```diff
@@ -45,12 +45,27 @@ PIPELINE_PCM_ADD(sof/pipe-low-latency-capture.m4,
...
+
 # PCM Media Playback pipeline 3 on PCM 1 using max 2 channels of s32le.
 # 2000us deadline on core 0 with priority 1
 PIPELINE_PCM_ADD(sof/pipe-pcm-media.m4,
 	3, 1, 2, s32le,
-	2000, 1, 0,
-	8000, 96000, 48000)
+	1000, 0, 0,
+	8000, 96000, 48000,
+	0, PIPELINE_PLAYBACK_SCHED_COMP_1)
```
Comment (plbossart) [inline tools/topology/sof-bdw-codec.m4]: why is this different from pipeline 1 in terms of scheduling?

File: tools/topology/CMakeLists.txt [hunk 2]
```diff
@@ -19,6 +19,8 @@ add_custom_target(abi
 set(TPLGS
 	"sof-cht-nocodec\;sof-cht-nocodec\;-DPLATFORM=cht"
 	"sof-cht-nocodec\;sof-byt-nocodec\;-DPLATFORM=byt"
+	"sof-hsw-nocodec\;sof-hsw-nocodec\;-DPLATFORM=hsw"
```
Comment (plbossart) [inline tools/topology/CMakeLists.txt]: we don't support haswell, why bother?

File: tools/topology/sof-hsw-nocodec.m4 [hunk 3]
```diff
@@ -0,0 +1,97 @@
+#
+ifelse(PLATFORM, `hsw', `# Topology for generic CherryTrail board with no codec.', `')
+ifelse(PLATFORM, `bdw', `# Topology for generic BayTrail board with no codec.', `')
```
Comment (plbossart) [inline tools/topology/sof-hsw-nocodec.m4]: What do Baytrail and cherrytrail do here?

File: tools/topology/sof-hsw-nocodec.m4 [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+include(`common/tlv.m4')
+
+# Include Token library
+include(`sof/tokens.m4')
+
+# Include DSP configuration
+ifelse(PLATFORM, `hsw', include(`platform/intel/hsw.m4'),
+	ifelse(PLATFORM, `bdw', include(`platform/intel/bdw.m4'), `'))
+define(PIPE_NAME, ifelse(PLATFORM, `hsw', pipe-hsw-nocodec,
+	ifelse(PLATFORM, `bdw', pipe-bdw-nocodec, `')))
```
Comment (plbossart) [inline tools/topology/sof-hsw-nocodec.m4]: forget hsw

File: tools/topology/sof-hsw-nocodec.m4 [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	PIPELINE_SOURCE_1, 2, s24le,
+	1000, 0, 0)
+
+# PCM Media Playback pipeline 3 on PCM 1 using max 2 channels of s32le.
+# 1000us deadline on core 0 with priority 1
+PIPELINE_PCM_ADD(sof/pipe-pcm-media.m4,
+	3, 1, 2, s32le,
+	1000, 1, 0,
+	8000, 48000, 48000,
+	0, PIPELINE_PLAYBACK_SCHED_COMP_1)
```
Comment (plbossart) [inline tools/topology/sof-hsw-nocodec.m4]: why is this only here?

File: tools/topology/sof-bdw-codec.m4 [hunk 6]
```diff
@@ -45,12 +45,27 @@ PIPELINE_PCM_ADD(sof/pipe-low-latency-capture.m4,
...
+# SSP port 0 is our only pipeline DAI
+#
+
+# playback DAI is SSP0 using 2 periods
+# Buffers use s24le format, 1000us deadline on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+	1, SSP, 0, Codec,
+	PIPELINE_SOURCE_1, 2, s24le,
+	1000, 0, 0)
+
```
[the remark this answers is not in the document]
Comment (xiulipan) [reply tools/topology/sof-bdw-codec.m4]: @plbossart copy paste from the working sof-byt-codec.m4

File: tools/topology/sof-hsw-nocodec.m4 (hunk 3 above)
Comment (xiulipan) [reply tools/topology/sof-hsw-nocodec.m4]: @plbossart copy paste from the working sof-byt-nocodec.m4, will fix.

File: tools/topology/sof-bdw-codec.m4 (hunk 1 above)
Comment (xiulipan) [reply tools/topology/sof-bdw-codec.m4]: not sure, I follow the working m4 from cht. I think they all use the same DMA and maybe should follow the same schedule.

File: tools/topology/CMakeLists.txt (hunk 2 above)
Comment (xiulipan) [reply tools/topology/CMakeLists.txt]: OK, will remove. So do we need to remove the build for HSW FW as well?

**Comment by xiulipan** [comment]
@plbossart updated the nocodec part.
@ranj063 Why we have the dai before media in byt nocodec?
@tlauda Do we need to change the scheduler for BYT/BDW ?

**Comment by ranj063** [comment]
> @ranj063 Why we have the dai before media in byt nocodec?

@xiulipan because the media pipeline needs to know the scheduling comp (PIPELINE_PLAYBACK_SCHED_COMP_1) and that is set by DAI_ADD() for pipeline 1

File: tools/topology/CMakeLists.txt (hunk 2 above)
Comment (plbossart) [reply tools/topology/CMakeLists.txt]: yes

**Comment by zrombel** [comment]
Due to PR#2029 merge and CI bit depth conversion tests changes, this PR require rebase.

**Comment by plbossart** [comment]
@xiulipan can you rebase and resubmit?

**Comment by xiulipan** [comment]
@plbossart Sure, will try to fix the BDW FW and rebase tplg first.

**Comment by xiulipan** [comment]
@plbossart @jajanusz @ranj063 update with bdw-rt5677 topology.
Test with https://github.com/thesofproject/linux/pull/1484 fw and tplg load is good.
Playback is also work with https://github.com/thesofproject/sof/pull/2188

File: tools/topology/sof-bdw-codec.m4 [hunk 7]
```diff
@@ -96,3 +95,13 @@ DAI_CONFIG(SSP, 0, 0, Codec,
 		      SSP_CLOCK(fsync, 48000, codec_slave),
 		      SSP_TDM(2, 25, 3, 3),
 		      SSP_CONFIG_DATA(SSP, 0, 24)))
+
+dnl CODEC is defined and will be expanded, need to undefine it before use
+
+ifelse(CODEC, `RT5677',
```
Comment (ranj063) [inline tools/topology/sof-bdw-codec.m4]: @xiulipan there is no need to make the virtual widgets conditional at all. As long as they have unique IDs, no harm in just adding all of them

**Comment by ranj063** [comment]
fix typo in commit message "Add virtual widget"

File: tools/topology/sof-bdw-codec.m4 (hunk 7 above)
Comment (xiulipan) [reply tools/topology/sof-bdw-codec.m4]: @ranj063 Even if there is no need to add that virtual widgets?

**Comment by lgirdwood** [comment]
CI usual suspects

**Pull request merged**